### PR TITLE
Update bitnami/postgresql dependency chart to latest version

### LINKED ISSUE #27353: Upgrade to the latest version of bitnami/postgresql
### Official Helm Chart version

1.7.0 (latest released)

This is a feature request for the helm chart. It's currently using bitnami/postgresql as a dependency chart when Postgres is enabled. It's pointing at version 10.5.3, which is over a year old. That chart has been updated significantly since then. It offers a lot of new features, and also has a lot of changes in variable names. This can make it difficult to customize values of that chart since you have to find the documentation for it at that version.

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by calebwoofenden**
Update from version 10.5.3 to version 12.1.2 of the bitnami/postgresql chart. This version changes the name of the username and password variables, so I've changed those here as well. Version 10.5.3 is over a year old and a lot of new features have been added since then.

I also updated the dependency for bitnami/postgresql listed in the `Chart.yaml` to point at bitnami's chart repo. This means you can just change the version of the chart in `Chart.yaml` and then run `helm dependency update` and it will fetch that version for you and place the `.tgz` file in the `charts` dir and update the `Chart.lock`

Closes: https://github.com/apache/airflow/issues/27353

### CODE REVIEW

**Comment by potiuk** [comment]
Oh yeah. Some test failures to fix :). Good that we have them. But they should be clear guidelines on what to fix :)

File: chart/Chart.lock
Comment (potiuk) [inline chart/Chart.lock]: Do I understand it correctly that we do not HAVE TO update to get it works (I see that the binary .tgz is replaced not removed)

If so - this is really cool.

We had it initially done like that but that caused unnecessary dependency on the external chart which is a bad idea in some scenarios - for example when user instals the charts behind firewalls from local sources. It also introduced instability in our CI process. 

So we decided to embed the .tgz (see https://github.com/apache/airflow/issues/24037 and PR #24395)

However - if I understand correctly - this one is purely to make it easier in the future if we want to upgrade (or it makes it easy for users to upgrade if they want to them themselves? This does not introduce "necessity" of running `helm update` correct?

**Comment by potiuk** [comment]
I also rebased to latest `main` - partially to help you to see the errors better and partially to verify two improvements I've done to surface "helm test" errors better and to fix the "noise" generated by static checks (side-effect of another change I implemented recently). I hope it will be clearer now to see the errors that should be fixed @calebwoofenden :)

File: chart/Chart.lock
Comment (calebwoofenden) [reply chart/Chart.lock]: I see, thanks for the explanation. I was wondering why the whole .tgz was included, but that makes sense.

> However - if I understand correctly - this one is purely to make it easier in the future if we want to upgrade (or it makes it easy for users to upgrade if they want to them themselves? This does not introduce "necessity" of running helm update correct?

Yes, that's correct. Since the .tgz is still included, people shouldn't need to run `helm dependency update` just to install the chart. This should just mean that all it will take to update the version of the bitnami/postgresql chart used here is to update the version in the `Chart.yaml` and then run `helm dependency update` and it should pull in the new chart .tgz and update the `Chart.lock`.

Comment (potiuk) [reply chart/Chart.lock]: Perfect! Thanks for doing it this way :)

**Comment by calebwoofenden** [comment]
> Oh yeah. Some test failures to fix :). Good that we have them. But they should be clear guidelines on what to fix :)

I see one about the newsfragment file which is an easy change. I also see this: 
```
Error:  templates/: values don't meet the specifications of the schema(s) in the following chart(s):
airflow:
- postgresql.auth: Additional property enablePostgresUser is not allowed
- postgresql.auth: Additional property existingSecret is not allowed
- postgresql.auth: Additional property postgresPassword is not allowed
- postgresql.auth: Additional property usePasswordFiles is not allowed
- postgresql.auth: Additional property database is not allowed
- postgresql.auth: Additional property replicationPassword is not allowed
- postgresql.auth: Additional property replicationUsername is not allowed
- postgresql.auth: Additional property secretKeys is not allowed
```
I guess because I'm including a couple of properties under `postgresql.auth` it wants me to include all of the properties that are available under that key from the bitnami chart?

**Comment by jedcunningham** [comment]
Probably better to just turn `additionalProperties` on instead.

**Comment by calebwoofenden** [comment]
Looks like tests are passing now! The new version of the postgres chart changes the "headless" service name suffix from "headless" to "hl" and that was throwing off a bunch of tests. Thanks for the tip about the `additionalProperties` parameter also.

File: chart/Chart.lock
Comment (jedcunningham) [inline chart/Chart.lock]:
```suggestion
  version: 12.0.1
```

This version was just released, might as well jump to it since we aren't quite ready to merge.

File: chart/Chart.yaml
Comment (jedcunningham) [inline chart/Chart.yaml]:
```suggestion
    version: 12.0.1
```

File: newsfragments/27353.significant.rst
Comment (jedcunningham) [inline newsfragments/27353.significant.rst]:
```suggestion
The helm chart is now using a newer version of bitnami/postgresql (from 10.5.3 to 12.0.1)
```

**Review by jedcunningham — COMMENTED** [review]
@calebwoofenden, can we also [fail](https://masterminds.github.io/sprig/flow_control.html) the install via NOTES if folks are using the old user/password config location?

In theory no one is using this for real instances, but I'd rather be defensive here.

(Also, I realize this is a breaking change, but I think it's worthwhile given this should be a dev only scenario)

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
I've also verified the tarball is the same as the one from bitnami.

File: chart/templates/NOTES.txt
Comment (jedcunningham) [inline chart/templates/NOTES.txt]:
```suggestion
   {{ fail "postgresql.postgresqlUsername and postgresql.postgresqlPassword have been renamed to postgresql.auth.username and postgres.auth.password" }}
```

Looks like this is causing the unit test failures. Can you test this part out manually as well? Thanks.

**Review by jedcunningham — CHANGES_REQUESTED** [review]

File: chart/templates/NOTES.txt
Comment (calebwoofenden) [reply chart/templates/NOTES.txt]: Ah, good catch. Fixed the missing end quote. I just tested locally by running this command from the root of the repo: 
```
helm install --debug --dry-run airflow ./chart --set-string postgresql.postgresqlPassword=blah
```
and same with `postgresql.postgresqlUsername` and it failed as expected in both situations. It succeeded without the `--set-string`.

**Review by jedcunningham — APPROVED** [review]

**Comment by potiuk** [comment]
> Not really sure what to make of the test failures now, though I think I saw this failure on a few previous test runs (but not all).
> 
> ```
> >       assert sorted(list_of_sa_names_in_objects) == sorted(CUSTOM_SERVICE_ACCOUNT_NAMES)
> E       AssertionError: assert ['TestCleanup...stRedis', ...] == ['TestCleanup...stRedis', ...]
> E         Left contains one more item: 'default'
> E         Use -v to get the full diff
> ```
> 
> It seems like this is saying there is an extra ServiceAccount being created called `default`?

You can just run the test yourself. It's easy to run the tests and you should be able to debug it with pytest. You can run the pytest tests inside `breeze` development environment - https://github.com/apache/airflow/blob/main/BREEZE.rst. Should take few minutes to set-up and then once you enter Breeze you shouls be able to run `pytest tests/chart/.....` and debug and see more details on what is the problem.

Also apparently the K8S tests fail - airflow installation via the Helm Chart fails with timeout. You can a https://github.com/apache/airflow/blob/main/TESTING.rst#running-tests-with-kubernetes - there you can see even step-by-step instructions how to create cluster, deploy airflow and run tests. Likely you would need to see in the k8s logs why installing airflow via the Helm chart fails (again breeze has full support to make it super- easy including running k9s tool that helps with k8s inspection and debugging).

**Comment by calebwoofenden** [comment]
Hm, I merged main again a couple times and now the tests are passing. Could it be flaky tests?

**Comment by potiuk** [comment]
You are a new user and you have not merged any PR to Airlfow yet, so by github limitations we need to approve your workflows after you rebase/merge. I just did, so we will yet see if they fail

**Comment by potiuk** [comment]
But no - I believe they were not flakes

**Comment by potiuk** [comment]
Yep. Sounds like the problem is real.

**Comment by calebwoofenden** [comment]
Thanks, I didn't realize that the full tests were not running unless you triggered them. I ran the Breeze setup locally and identified the issue.

The `postgres` user was being created with a randomly-generated password instead of using the value in `auth.password`. The parameters of the bitnami/postgresql chart didn't behave exactly how I thought - it turns out that `auth.username` and `auth.password` are only for accounts other than the `postgres` admin user. To set the password for the `postgres` user you need to set `auth.postgresPassword`. I added this parameter as well as `auth.enablePostgresUser`, and added more detail to the `NOTES.txt` fail message.

**Comment by calebwoofenden** [comment]
Fixed the json schema for the new params to allow null. There's still one test failure, `test_rbac.py:test_service_account_custom_names_in_objects`. It seems to be failing because the namespace includes a ServiceAccount called `default`, which it seemingly doesn't expect. As far as I can tell every namespace will contain a `default` ServiceAccount. I'm kind of at a loss on this one.

**Comment by potiuk** [comment]
> Thanks, I didn't realize that the full tests were not running unless you triggered them. I ran the Breeze setup locally and identified the issue.

Only for new contributor who never merged a PR before :) . Once we merge this one, the word (Airlfow's one) is yours :)

**Comment by potiuk** [comment]
> Fixed the json schema for the new params to allow null. There's still one test failure, test_rbac.py:test_service_account_custom_names_in_objects. It seems to be failing because the namespace includes a ServiceAccount called default, which it seemingly doesn't expect. As far as I can tell every namespace will contain a default ServiceAccount. I'm kind of at a loss on this one.

Let's see the results

**Comment by potiuk** [comment]
i re-run the tests for sqlite, but seems like the Unit tests for Helm Chart need to be fixed.

**Comment by potiuk** [comment]
> i re-run the tests for sqlite, but seems like the Unit tests for Helm Chart need to be fixed.

Yep. one test failing.

**Comment by calebwoofenden** [comment]
Yes, it looks like it's the same failure I saw with this test case finding a ServiceAccount called `default` and not liking that. Every namespace must have a `default` ServiceAccount though. I don't really know where to start with this one.

**Comment by potiuk** [comment]
I just re-run it let's see. I would have to see the tests to know where it came from.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Needs fixes I am afraid.

**Comment by potiuk** [comment]
I think there is one simple test to fix (and I guess we are close to release a Helm chart shortly (we just released Airflow 2.5.0 and usually Helm Chart follows shortly).  Can you please fix the problem @calebwoofenden ?

**Comment by dnskr** [comment]
Hi @calebwoofenden!
It would be great to have your PR merged to 1.8.0 chart release. Will you have time to have a look and fix the test? Looks like it should be easy to make the last step after great effort you made.

**Comment by potiuk** [comment]
> Hi @calebwoofenden! It would be great to have your PR merged to 1.8.0 chart release. Will you have time to have a look and fix the test? Looks like it should be easy to make the last step after great effort you made.

Yep. Would be great to merge that one in :)

**Comment by snjypl** [comment]
@potiuk  i have opened a PR with fixes to the failing unittest and also upgraded the chart to lastest version 12.1.9 

https://github.com/apache/airflow/pull/29071 

i dont have permission to update this PR with those changes.

**Comment by potiuk** [comment]
Closed as #29071 has been merged.

**Pull request closed without merging**
